src/libs/decoders/vorbis.c

```diff
@@ -0,0 +1,238 @@
+        case VORBIS_bad_packet_type: return "VORBIS: bad packet type";
+        case VORBIS_cant_find_last_page: return "VORBIS: can't find last page";
+        case VORBIS_seek_failed: return "VORBIS: seek failed";
+        case VORBIS_ogg_skeleton_not_supported: return "VORBIS: multi-track streams are not supported; "
+                                                       "consider re-encoding without the Ogg Skeleton bitstream";
+        default: break;
+    } /* switch */
+
+    return "VORBIS: unknown error";
+} /* vorbis_error_string */
+#endif
+
+static int VORBIS_init(void)
+{
+    return 1;  /* always succeeds. */
+} /* VORBIS_init */
+
+static void VORBIS_quit(void)
+{
+    /* it's a no-op. */
+} /* VORBIS_quit */
+
+static int VORBIS_open(Sound_Sample *sample, const char *ext)
+{
+    Sound_SampleInternal *internal = (Sound_SampleInternal *) sample->opaque;
+    SDL_RWops *rw = internal->rw;
+    int err = 0;
+    stb_vorbis *stb = stb_vorbis_open_rwops(rw, 0, &err, NULL);
+
+	if(stb == NULL) {
+		SNDDBG(("%s (error code: %d)\n", vorbis_error_string(err), err));
+        return 0;
+	}
+    internal->decoder_private = stb;
+    sample->flags = SOUND_SAMPLEFLAG_CANSEEK;
+    sample->actual.format = AUDIO_S16SYS; // returns byte-order native to the running architecture
+    sample->actual.channels = stb->channels;
+    sample->actual.rate = stb->sample_rate;
+    const unsigned int num_frames = stb_vorbis_stream_length_in_samples(stb);
+    if (!num_frames)
+        internal->total_time = -1;
+    else
+    {
+        const unsigned int rate = stb->sample_rate;
+        internal->total_time = (num_frames / rate) * 1000;
+        internal->total_time += (num_frames % rate) * 1000 / rate;
+    } /* else */
+
+    return 1; /* we'll handle this data. */
+} /* VORBIS_open */
+
+
+static void VORBIS_close(Sound_Sample *sample)
+{
+    Sound_SampleInternal *internal = (Sound_SampleInternal *) sample->opaque;
+    stb_vorbis *stb = (stb_vorbis *) internal->decoder_private;
+    stb_vorbis_close(stb);
+} /* VORBIS_close */
+
+
+static Uint32 VORBIS_read(Sound_Sample *sample)
+{
+    Uint32 retval;
+    int rc;
+    int err;
+    Sound_SampleInternal *internal = (Sound_SampleInternal *) sample->opaque;
+    stb_vorbis *stb = (stb_vorbis *) internal->decoder_private;
+    const int channels = (int) sample->actual.channels;
+    const int want_samples = (int) (internal->buffer_size / sizeof (int16_t));
+
+    stb_vorbis_get_error(stb);  /* clear any error state */
+    rc = stb_vorbis_get_samples_short_interleaved(stb, channels, (int16_t *) internal->buffer, want_samples);
+    retval = (Uint32) (rc * channels * sizeof (int16_t));  /* rc == number of sample frames read */
+    err = stb_vorbis_get_error(stb);
+
+    if (retval == 0)
+    {
+        if (!err)
+            sample->flags |= SOUND_SAMPLEFLAG_EOF;
+        else
+        {
+            sample->flags |= SOUND_SAMPLEFLAG_ERROR;
+        } /* else */
```

Comment:
I would rewrite this statement as:

    sample->flags |= (err ? SOUND_SAMPLEFLAG_ERROR : SOUND_SAMPLEFLAG_EOF);

Reply:
Reads much nicer now.
``` c
    if (retval == 0) {
        sample->flags |= (err ? SOUND_SAMPLEFLAG_ERROR : SOUND_SAMPLEFLAG_EOF);
    }
    else if (retval < internal->buffer_size) {
        sample->flags |= SOUND_SAMPLEFLAG_EAGAIN;
    }
```

Reply:
Fixed in 2c9713227fcb0bc32a0f953050da0ea3111373b7